decision-records/006-javascript-compatibility.md

```diff
@@ -0,0 +1,161 @@
+
+### Import maps
+
+Import maps could provide another way to only load GOV.UK Frontend in browsers with a certain level of support for modern JavaScript features. It's [not widely supported enough yet](https://caniuse.com/import-maps), but could be a good consideration if we revisit which baseline we want to have in the future.
+
+## Implications
+
+### Automating parts of the development process
+
+When writing or changing JavaScript we currently have to remember not to use features that aren't available in the oldest browsers we support, or to manually add a polyfill. We also rely heavily on manual testing across multiple browsers.
+
+Improving our build process by introducing transpilation and additional linting rules:
+
+- improves the developer experience, as we provide additional safety guards against using unsupported features
+- enables us to use a modern syntax, including future features as long as our tooling can transpile it
+
+At the same time, we need to recognise that we are deferring some of the responsibility of ensuring browser compatibility to our linting and transpilation tools.
+
+If we find that the transpilation is not producing code that works in the targeted browsers, we may be relying on the maintainers of those dependencies to help us address issues.
+
+We will also need to carefully monitor the impact of transpilation as the code we ship will potentially be further removed from the code we author.
+
+### Improving the experience on older devices
+
+Restricting JavaScript to running only on browsers that support `type="module"` means that it will not run on the oldest browsers, which are likely to be on old, underpowered devices. This means that users on older devices will get a simplified, but likely more performant experience.
+
+### Pins the JavaScript we ship to 2017
+
+This approach pins us to shipping code based on browsers that shipped in 2017.
+
+With the exception of Internet Explorer 11 (for which we've decided to not provide JavaScript enhancements for), [this target](https://browsersl.ist/#q=supports+es6-module) covers further than [the 95%](https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices#understanding-the-table) of [visits to GOV.UK (internal link)](https://docs.google.com/presentation/d/17gcoKyQy6YnfcBG6LGIkNKi6VVOz9nc_EffvA39o98Y/edit#slide=id.g1bee1a35b37_0_83).
+
+The target creates a long tail of browsers that users no longer use and for which we'll ship code transformations. This long tail will only expand as browser vendors ship newer versions.
+
+We need to be mindful that:
+
+- some of the transformations may add extra weight to the JS that are only required for a tiny percentage of users
```

Comment:
Could we make this a tiny bit more obvious?

Assuming we're talking about the trade off in KB file size versus just guarding an optional feature

```suggestion
- some of the transformations may add extra file size to the JS to provide compatibility to only a tiny percentage of users
```